main/lsp/LSPTypechecker.h

```diff
@@ -27,23 +31,44 @@ class TypecheckRun final {
     LSPFileUpdates updates;
     // Specifies if the typecheck run took the fast or slow path.
     bool tookFastPath = false;
-    // Specifies if the typecheck run was canceled.
-    bool canceled = false;
     // If update took the slow path, contains a new global state that should be used moving forward.
     std::optional<std::unique_ptr<core::GlobalState>> newGS;
 
     TypecheckRun(std::vector<std::unique_ptr<core::Error>> errors = {},
                  std::vector<core::FileRef> filesTypechecked = {}, LSPFileUpdates updates = {},
                  bool tookFastPath = false, std::optional<std::unique_ptr<core::GlobalState>> newGS = std::nullopt);
-
-    // Make a canceled TypecheckRun.
-    static TypecheckRun makeCanceled();
 };
 
 /**
  * Encapsulates typechecker operations and enforces that they happen on a single thread.
  */
 class LSPTypechecker final {
+    /**
+     * Contains the LSPTypechecker state that is needed to cancel a running slow path operation and any subsequent fast
+     * path operations that have preempted it.
+     */
+    class UndoState final {
```

Comment:
this has a lot of templated classes.
Can we separate it into a separate header for compilation performance reason and have this file take it forward declared?

Reply:
Sure, I can do that.